extensions/amp-date-countdown/0.1/amp-date-countdown.js

```diff
@@ -0,0 +1,261 @@
+
+    /** @private {string} */
+    this.biggestUnit_ = '';
+
+    /** @const {!../../../src/service/template-impl.Templates} */
+    this.templates_ = Services.templatesFor(this.win);
+  }
+
+  /** @override */
+  buildCallback() {
+    const self = this;
+    //Note: One of datetime, timestamp-ms, timestamp-seconds is required.
+    this.endDate_ = this.element.getAttribute('end-date');
+    this.timestampMs_ = Number(this.element.getAttribute('timestamp-ms'));
+    this.timestampSeconds_
+      = Number(this.element.getAttribute('timestamp-seconds'));
+    this.offsetSeconds_ = Number(this.element.getAttribute('offset-seconds'));
+
+    const locale = this.element.getAttribute('locale') || 'en';
+    const whenEnded = this.element.getAttribute('when-ended') || 'stop';
+    this.biggestUnit_ = (this.element.getAttribute('biggest-unit')) || 'days';
+
+    let epoch = this.getEpoch_(); //1519776000000
+    if (this.offsetSeconds_) {
+      epoch += this.offsetSeconds_ * 1000;
+    }
+    let differentBetween = new Date(epoch) - new Date();
+    let data;
+    const localeWordList = this.getLocaleWord_(locale);
+    const delay = 1000;
+
+    const countDownTicker = setInterval(function() {
+      //console.log('differentBetween', differentBetween);
+      data = self.getYDHMSFromMs_(differentBetween);
+      if (whenEnded === 'stop' && differentBetween < 1000) {
+        Services.actionServiceForDoc(self.element)
+            .trigger(self.element, 'timeout', null, ActionTrust.HIGH);
+        clearInterval(countDownTicker);
+      }
+      differentBetween -= delay;
+      self.renderItems_(Object.assign(data, localeWordList));
+    }, delay);
+
+    this.registerAction('timeout', this.timeout_.bind(this));
+
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /**
+   * @param {!Object} items
+   * @return {!Promise}
+   * @private
+   */
+  renderItems_(items) {
+    return this.templates_.findAndRenderTemplate(this.element, items)
+        .then(this.boundRendered_);
+  }
+
+  /**
+   * @return {number}
+   * @private
+   */
+  getEpoch_() {
+    let epoch;
+
+    if (this.endDate_) {
+      epoch = Date.parse(this.endDate_);
+    } else if (this.timestampMs_) {
+      epoch = this.timestampMs_;
+    } else if (this.timestampSeconds_) {
+      epoch = this.timestampSeconds_ * 1000;
+    }
+
+    if (isNaN(epoch)) {
+      user().error(TAG,
+          'One of end-date, timestamp-ms, timestamp-seconds is required');
+    }
+
+    //console.log('epoch', epoch);
+    return epoch;
+  }
+
+  /**
+   * Toggles the timeout_ state when timer hit 0.
+   * @param {?../../../src/service/action-impl.ActionInvocation=} opt_invocation
+   * @private
+   */
+  timeout_() {
+
+  }
+
+  /**
+   * @param {string} locale
+   * @return {Object}
+   * @private
+   */
+  getLocaleWord_(locale) {
+    //https://ctrlq.org/code/19899-google-translate-languages refer to google code
+    const localeWord = {
```

Comment:
You should make this object a file-level variable, and make the values into arrays so the code can be faster without splitting the string every time the function is called.

Reply:
I have changed them into Array format, but i dont really get what do you mean by "file-level" variable, putting entire `localeWord` into **third_party** path and then import them as variable into this `amp-date-countdown.js` file ?

Reply:
By file-level variable I mean outside this class. So line 27 of this file would be a good place to move this to